---
title: "R Assignment"
author: "Kimia Noroozi"
date: "2023-03-10"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



#Data Inspection
```{r}
# Load files
# Read fang_et_al_genotypes.txt 
genotypes <- read.delim(file="fang_et_al_genotypes.txt", sep="\t", header=TRUE)

# Read snp_position.txt
snp <- read.delim(file="snp_position.txt", sep="\t",header= TRUE)

## Data inspection for fang_et_al_genotypes.txt


type_of_genotypes <- typeof(genotypes)
# genotypes is a "list"

length_of_genotypes <- length(genotypes)
# length of this list is 986

number_of_rows_genotypes <- dim(genotypes)[1]
number_of_columns_genotyps <- dim(genotypes)[2]
# dimensions of this file includes: 2782 rows and 986 columns

info_genotypes <- str(genotypes)
# this file is a 'data.frame' that includes 2782 observations of 986 variables, including Sample_ID, JG_OTU,Group, and genotypes.

class_genotypes <- sapply(genotypes,class)
# all columns are "characters"

number_of_groups_genotypes <- table(genotypes$Group)
# number of inputs for our desired groups in the file are: ZMMIL:290, ZMMLR:1256, ZMMMR:27, ZMPBA: 900, ZMPIL:40, ZMPJA:34 

## Data inspection for snp_position.txt
type_of_file_snp <- typeof(snp)
# snp is a "list"

length_of_file_snp <- length(snp)
# length of this list is 15

number_of_rows_snp <- dim(snp)[1]
number_of_columns_snp <- dim(snp)[2]
# dimensions of this file includes: 983 rows and 15 columns

info_snp <- str(snp)
# this file is a 'data.frame' that includes 983 observations of 15 variables, including SNP_ID, chromosome, position, etc.

class_snp <- sapply(snp,class)
# 9 columns are "characters" and 6 are "integers"
```



#Data Procesing
```{r}
library(tidyverse)
library(dplyr)


# Subset Maize information (ZMMIL,ZMMLR,ZMMMR)
maize <- filter(genotypes, Group %in% c("ZMMIL","ZMMLR","ZMMMR"))

# Subset Teosinte information (ZMPBA,ZMPIL,ZMPJA)
teosinte <- filter(genotypes, Group %in% c("ZMPBA","ZMPIL","ZMPJA"))

# Remove unnecessary columns (columns 1-3 for genotypes, columns 2 and 5:15 for snp)
maize_cut <- maize[-c(1:3)]
teosinte_cut <- teosinte[-c(1:3)]
snp <- snp[c(1,3,4)]


# Transpose genotype files to
maize_t <- t(maize_cut)
teosinte_t <- t(teosinte_cut)

# Merge genotype files with snp position file
maize_snp <-  merge(snp,maize_t,by.x="SNP_ID", by.y="row.names")
teosinte_snp <-  merge(snp,teosinte_t,by.x = "SNP_ID",by.y = "row.names")

dir.create("maize")
dir.create("teosinte")

#sorting of maize
# for numbers from 1 to 10:
for (i in 1:10){
  #filter all occurrences of i in maize_snp based on the chromosome column,
  maize.c <- filter(maize_snp,Chromosome %in% c(i))
  
  #sort the generated file based on the Position column in increasing order, 
  maize.csi <- maize.c[order(maize.c$Position),]
  
  #create corresponding files with the correct name based on increasing values and i,
  m1 <- paste("maize/maize_inc",i,".txt", sep="")
  
  #write the results of search and sort in the file created with the corresponding name.
  write.table(maize.csi,file= m1, row.names = FALSE, sep = "\t")
  
  #sort the generated file again based on the Position column in decreasing order,
  maize.csdf <- maize.c[order(maize.c$Position,decreasing = TRUE),]
  
  #replace ? values with -
  maize.csdf[maize.csdf=="?/?"] <- "-/-"
  
  #create corresponding files with the correct name based on decreasing values and i,
  m2 <- paste("maize/maize_dec",i,".txt",sep="")
  
  #write the results of searcha nd sort in the file created with the corresponding name.
  write.table(maize.csdf,file=m2,row.names = FALSE, sep="\t")
}


#sorting for teosinte
# for numbers from 1 to 10:
for (i in 1:10){
  #filter all occurances of i in maize_snp based on the chromosome column,
  teosinte.c <- filter(teosinte_snp,Chromosome %in% c(i))
  
  #sort the generated file based on the Position column in increasing order, 
  teosinte.csi <- teosinte.c[order(teosinte.c$Position),]
  
  #create corresponding files with the correct name based on increasing values and i,
  t1 <- paste("teosinte/teosinte_inc",i,".txt", sep="")
  
  #write the results of search and sort in the file created with the corresponding name.
  write.table(teosinte.csi,file= t1, row.names = FALSE, sep = "\t")
  
  #sort the generated file again based on the Position column in decreasing order,
  teosinte.csd <- teosinte.c[order(teosinte.c$Position,decreasing = TRUE),]
  
  #replace ? values with -,
  teosinte.csd[teosinte.csd=="?/?"] <- "-/-"
  
  #create corresponding files with the correct name based on decreasing values and i,
  t2 <- paste("teosinte/teosinte_dec",i,".txt",sep="")
  
  #write the results of search a nd sort in the file created with the corresponding name.
  write.table(teosinte.csd,file=t2,row.names = FALSE, sep="\t")
}
```
  
#Data Visualization

```{r}
library(ggplot2)

#create new data frames for the plots that include both mazie and teosinte groups 
mt <- filter(genotypes,Group %in% c("ZMMIL","ZMMLR","ZMMMR","ZMPBA","ZMPIL","ZMPJA"))
mt <- mt[-c(1,2)]
mt <- t(mt)
mt_snp <- merge(snp,mt,by.x = "SNP_ID", by.y = "row.names")

#plot SNPs per chromosome
ggplot (data = mt_snp) + geom_bar(mapping = aes(x = Chromosome, fill = Chromosome)) +xlab(label = "Chromosome") + ylab(label = "SNPs") + ggtitle("SNPs per Chromosome") + scale_x_discrete(limit=c(1:10, "unknown", "multiple"))
## Chromosome 1 has the most SNPs

#plot SNPs per chromosome in each group (maize vs. teosinte)

##maize
ggplot (data = maize_snp) + geom_bar(mapping = aes(x = Chromosome, fill = Chromosome)) +xlab(label = "Chromosome") + ylab(label = "SNPs") + ggtitle("Maize: SNPs per Chromosome") + scale_x_discrete(limit=c(1:10, "unknown", "multiple"))

##teosinte
ggplot (data = teosinte_snp) + geom_bar(mapping = aes(x = Chromosome, fill = Chromosome)) +xlab(label = "Chromosome") + ylab(label = "SNPs") + ggtitle("Teosinte: SNPs per Chromosome") + scale_x_discrete(limit=c(1:10, "unknown", "multiple"))


#for homozygousity and heterozygosity

##for maize
# Loop through each row of the data frame
homogenety <- numeric(nrow(maize_snp))
missing_data <- numeric(nrow(maize_snp))

# Loop through each row of the data frame
for (i in 1:nrow(maize_snp)) {
  # Extract the row for each data as a vector
  row_i <- as.vector(maize_snp[i,])
  
  # Count the number of strings in the row that match any string in c("A/A", "C/C", "G/G", "T/T")
  count_i <- sum(row_i %in% c("A/A", "C/C", "G/G", "T/T"))
  
  # Store the result in the results vector
  homogenety[i] <- (count_i/(ncol(maize_snp)-3))
  count_mi <- sum(row_i %in% "?/?")
  
  # Store the result in the results vector
  missing_data[i] <- (count_mi/(ncol(maize_snp)-3))
}

maize_final <- cbind(maize_snp[,1], homogenety,missing_data, maize_snp[,2:ncol(maize_snp)])

# Set the name of the new column to the name of the vector
names(maize_final)[1] <- "SNP_ID"
names(maize_final)[2] <- "homogenety"
names(maize_final)[3] <- "missing_data"

ggplot(data = maize_final) + geom_bar(mapping = aes(x=homogenety))

##for teosinte

# Loop through each row of the data frame
homogenety_teo <- numeric(nrow(teosinte_snp))
missing_data_teo <- numeric(nrow(teosinte_snp))

# Loop through each row of the data frame
for (i in 1:nrow(teosinte_snp)) {
  # Extract the row for each data as a vector
  row_i <- as.vector(teosinte_snp[i,])
  
  # Count the number of strings in the row that match any string in c("A/A", "C/C", "G/G", "T/T")
  count_i <- sum(row_i %in% c("A/A", "C/C", "G/G", "T/T"))
  
  # Store the result in the results vector
  homogenety_teo[i] <- (count_i/(ncol(teosinte_snp)-3))
  count_mi <- sum(row_i %in% "?/?")
  
  # Store the result in the results vector
  missing_data_teo[i] <- (count_mi/(ncol(teosinte_snp)-3))
}

teosinte_final <- cbind(teosinte_snp[,1], homogenety,missing_data, teosinte_snp[,2:ncol(teosinte_snp)])

# Set the name of the new column to the name of the vector
names(teosinte_final)[1] <- "SNP_ID"
names(teosinte_final)[2] <- "homogenety"
names(teosinte_final)[3] <- "missing_data"

ggplot(data = teosinte_final) + geom_bar(mapping = aes(x=homogenety))


#Your own visualization
ggplot(data = teosinte_snp) + geom_point(mapping=aes(x=Position, y=Chromosome))
ggplot(data = maize_final) + geom_point(mapping = aes(x=Position, y=homogenety), alpha=0.1)


```
